quiz_app_main.rb

```diff
@@ -0,0 +1,208 @@
+# frozen_string_literal: true
+
+require_relative 'teacher'
+require_relative 'student'
+# class for controlling all flow
+class Main
+  def initialize
+    @storage = { 'Teacher' => [], 'Student' => [] }
+    @current_user = nil
+  end
+
+  private
+
+  ROLE_MAP = { 1 => Teacher, 2 => Student }
+
+  def signup_login_options
+    puts 'Choose an option: '
+    puts '1. Sign Up'
+    puts '2. Login'
+    puts '3. Exit'
+  end
+
+  def name_validation?(name)
+    if name.length <= 2 || name.length > 15
+      puts 'Re-enters correct name!(Should be 3 to 15 char)'
+      false
+    elsif !name.match?(/[a-zA-Z]/)
+      puts 'Name must be Alphabetic!'
+      false
+    else
+      true
+    end
+  end
+
+  def email_validation?(email)
+    if email.match?(/\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i)
+      true
+    else
+      puts 'Enter Valid Email (Gmail Account only)'
+      false
+    end
+  end
+
+  def password_validation?(password)
+    if password.length <= 8 || password.length > 15
+      puts 'Password should be 9 to 15 char'
+      false
+    else
+      true
+    end
+  end
+
+  def valid_email
+    email = ''
+    loop do
+      print 'Enter your Email: '
+      email = gets.strip
+      break if email_validation?(email)
+    end
+    email
+  end
+
+  def valid_password
+    password = ''
+    loop do
+      print 'Enter your Password: '
+      password = gets.strip
+      break if password_validation?(password)
+    end
+    password
+  end
+
+  def find_user_by_email(email)
+    @storage.each_value do |array|
+      user = array.find { |u| u.email == email }
+      return user if user
+    end
+    nil
+  end
```

Comment:
No need to return `nil`
BTW, we can enhance this method:
```suggestion
  def find_by(hash)
    key, value = hash.first
    @storage.values.flatten.find { |user| user.public_send(key) == value }
  end
```
and
`find_by(email: "mail@domain.com")`